File: SDKBridge/Android/bridge-plugin-example/src/main/java/com/sdkbridge/plugin/example/ExamplePlugin.java
package com.sdkbridge.plugin.example;

import android.app.Application;
import android.app.Activity;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.sdkbridge.core.plugin.BaseBridgePlugin;
import com.sdkbridge.core.util.JsonHelper;

/**
 * 示例插件 - 演示如何编写一个桥接插件
 * <p>
 * 这个示例展示了完整的插件结构，包括：
 * 1. 插件初始化
 * 2. 同步方法调用
 * 3. 异步方法调用（通过事件主动推送结果）
 * 实际接入第三方SDK时，照此模式替换具体SDK调用即可。
 * </p>
 */
public class ExamplePlugin extends BaseBridgePlugin {

    private static final String TAG = "ExamplePlugin";

    /** 插件名称 - Unity侧通过此名称调用 */
    private static final String PLUGIN_NAME = "Example";

    private boolean isInitialized = false;

    @NonNull
    @Override
    public String getPluginName() {
        return PLUGIN_NAME;
    }

    @Override
    public void onInit(@NonNull Activity activity, @Nullable Bundle params) {
        super.onInit(activity, params);

        // 从参数中读取配置（示例）
        String appKey = null;
        if (params != null) {
            String json = params.getString("json");
            appKey = JsonHelper.getString(json, "appKey");
        }

        Log.i(TAG, "ExamplePlugin 初始化, appKey=" + appKey);
        isInitialized = true;
    }

    @Nullable
    @Override
    public String execute(@NonNull String methodName, @Nullable String jsonParams) {
        Log.d(TAG, "执行方法: " + methodName + ", 参数: " + jsonParams);

        switch (methodName) {
            case "showToast":
                return handleShowToast(jsonParams);

            case "getDeviceInfo":
                return handleGetDeviceInfo();

            case "doAsyncWork":
                return handleAsyncWork(jsonParams);

            case "isInitialized":
                return successResult(String.valueOf(isInitialized));

            default:
                return errorResult(-1, "未知方法: " + methodName);
        }
    }

    // ==================== 方法实现示例 ====================

    /**
     * 同步方法示例 - 显示Toast
     */
    private String handleShowToast(@Nullable String jsonParams) {
        String message = JsonHelper.getString(jsonParams, "message");
        if (message == null) {
            return errorResult(1, "缺少message参数");
        }

        runOnUiThread(() -> Toast.makeText(activity, message, Toast.LENGTH_SHORT).show());
        return successResult(null);
    }

    /**
     * 同步方法示例 - 获取设备信息
     */
    private String handleGetDeviceInfo() {
        String info = "{" +
                "\"brand\":\"" + android.os.Build.BRAND + "\"," +
                "\"model\":\"" + android.os.Build.MODEL + "\"," +
                "\"sdk\":" + android.os.Build.VERSION.SDK_INT + "," +
                "\"version\":\"" + android.os.Build.VERSION.RELEASE + "\"" +
                "}";
        return successResult(info);
    }

    /**
     * 异步方法示例 - 模拟耗时操作后通过事件返回结果
     */
    private String handleAsyncWork(@Nullable String jsonParams) {
        int delayMs = JsonHelper.getInt(jsonParams, "delay", 2000);

        // 模拟异步操作
        new Thread(() -> {
            try {
                Thread.sleep(delayMs);
                // 异步完成后，通过事件推送结果到Unity
                sendEvent("onAsyncWorkDone", "{\"result\":\"异步任务完成\",\"timestamp\":" + System.currentTimeMillis() + "}");
            } catch (InterruptedException e) {
                sendEvent("onAsyncWorkDone", "{\"error\":\"任务被中断\"}");
            }
        }).start();

        // 同步返回"已接受"
        return successResult("{\"status\":\"accepted\"}");
    }

    // ==================== 生命周期 ====================

    @Override
    public void onApplicationCreate(@NonNull Application app) {
        Log.i(TAG, "ExamplePlugin onApplicationCreate called");
    }

    @Override
    public void onResume() {
        Log.d(TAG, "onResume");
        // 在这里恢复SDK状态（如广告SDK的恢复等）
    }

    @Override
    public void onPause() {
        Log.d(TAG, "onPause");
    }

    @Override
    public void onDestroy() {
        Log.d(TAG, "onDestroy");
        isInitialized = false;
        // 在这里释放SDK资源
    }
}

File: SDKBridge/Ios/ExamplePlugin.m
#import "ExamplePlugin.h"
#import "BridgeManager.h"
#import <UIKit/UIKit.h>
#import <UIKit/UIKit.h>

@interface ExamplePlugin ()
@property (nonatomic, strong) NSString *appKey;
@property (nonatomic, assign) BOOL isInitialized;
@end

@implementation ExamplePlugin

+ (void)load {
	//这里我启动就注册了插件，如果这个插件不需要启动就注册，也可以在unity侧寻找时机去注册
    [[BridgeManager sharedInstance] registerPlugin:NSStringFromClass(self) jsonParams:nil];
}

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    NSLog(@"[ExamplePlugin iOS] 拦截到 didFinishLaunchingWithOptions");
    return YES;
}

// 1. 实现 IBridgePlugin 的必选方法
#pragma mark - IBridgePlugin Protocol

- (NSString *)pluginName {
    return @"Example"; // 这个名字必须与 Unity 侧 Call("Example", ...) 的第一个参数一致
}

- (void)onInit:(NSDictionary *)params {
    self.isInitialized = YES;
    self.appKey = params[@"appKey"] ?: @"";
    
    NSLog(@"[ExamplePlugin iOS] 初始化完成. appKey = %@", self.appKey);
}

- (NSString *)execute:(NSString *)methodName params:(NSDictionary *)params {
    NSLog(@"[ExamplePlugin iOS] 执行方法: %@, 参数: %@", methodName, params);
    
    if ([methodName isEqualToString:@"getDeviceInfo"]) {
        return [self handleGetDeviceInfo];
    } 
    else if ([methodName isEqualToString:@"showToast"]) {
        return [self handleShowToast:params];
    } 
    else if ([methodName isEqualToString:@"doAsyncWork"]) {
        return [self handleAsyncWork:params];
    }
    
    // 对于未知方法
    return [self errorResultWithCode:-1 msg:[NSString stringWithFormat:@"未知方法: %@", methodName]];
}

// 2. 内部处理方法
#pragma mark - Method Handlers

/// 同步调用示例：获取设备信息
- (NSString *)handleGetDeviceInfo {
    NSString *brand = @"Apple";
    NSString *model = [[UIDevice currentDevice] model]; // 比如 "iPhone"
    NSString *systemVersion = [[UIDevice currentDevice] systemVersion];
    
    // 返回 JSON 格式的字符串
    NSString *info = [NSString stringWithFormat:@"{\"brand\":\"%@\",\"model\":\"%@\",\"version\":\"%@\"}", brand, model, systemVersion];
    
    return [self successResultWithData:info];
}

/// 同步调用示例：主线程操作 UI（在 iOS 中，我们用 UIAlertController 模拟 Toast）
- (NSString *)handleShowToast:(NSDictionary *)params {
    NSString *message = params[@"message"];
    if (!message || message.length == 0) {
        return [self errorResultWithCode:1 msg:@"缺少 message 参数"];
    }
    
    // UI 操作必须放在主线程执行
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:nil 
                                                                       message:message 
                                                                preferredStyle:UIAlertControllerStyleAlert];
        
        // 查找当前的根视图控制器
        UIViewController *rootVC = [UIApplication sharedApplication].keyWindow.rootViewController;
        if (rootVC) {
            [rootVC presentViewController:alert animated:YES completion:^{
                // 模拟 Toast 自动消失 (延迟1.5秒)
                dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                    [alert dismissViewControllerAnimated:YES completion:nil];
                });
            }];
        }
    });
    
    return [self successResultWithData:nil];
}

/// 异步调用示例：模拟耗时操作，并通过事件推送结果
- (NSString *)handleAsyncWork:(NSDictionary *)params {
    NSNumber *delayNumber = params[@"delay"];
    int delayMs = delayNumber ? [delayNumber intValue] : 2000;
    
    NSLog(@"[ExamplePlugin iOS] 发起异步任务，延迟: %d 毫秒", delayMs);
    
    // 模拟在后台线程执行异步任务
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        // 延时
        usleep(delayMs * 1000); 
        
        NSTimeInterval timestamp = [[NSDate date] timeIntervalSince1970] * 1000;
        NSString *eventData = [NSString stringWithFormat:@"{\"result\":\"异步任务完成(iOS)\",\"timestamp\":%f}", timestamp];
        
        // 耗时任务完成后，通过 BridgeManager 的 sendEvent 向 Unity 发送事件
        [[BridgeManager sharedInstance] sendEvent:[self pluginName] 
                                        eventName:@"onAsyncWorkDone" 
                                             data:eventData];
    });
    
    // 立即向 C# 同步返回一个成功，告诉 Unity "任务已接管"
    return [self successResultWithData:@"{\"status\":\"accepted\"}"];
}

// 3. 辅助方法
#pragma mark - Helper

- (NSString *)successResultWithData:(NSString *)data {
    if (data) {
        return [NSString stringWithFormat:@"{\"code\":0,\"msg\":\"success\",\"data\":%@}", data];
    }
    return @"{\"code\":0,\"msg\":\"success\"}";
}

- (NSString *)errorResultWithCode:(int)code msg:(NSString *)msg {
    return [NSString stringWithFormat:@"{\"code\":%d,\"msg\":\"%@\"}", code, msg];
}

@end
